pkg/operator/test/volumes.go

```diff
@@ -0,0 +1,245 @@
+}
+
+// HumanReadableVolume returns a string representation of a Kubernetes volume which is more compact
+// and human readable than the default string go prints.
+func HumanReadableVolume(v *v1.Volume) string {
+	sourceString := "something went wrong parsing the volume source in HumanReadableVolume()"
+	if v.VolumeSource.EmptyDir != nil {
+		sourceString = "EmptyDir"
+	} else if v.VolumeSource.HostPath != nil {
+		sourceString = v.VolumeSource.HostPath.Path
+	} else {
+		// If can't convert the vol source to something human readable, just output something useful
+		sourceString = fmt.Sprintf("%v", v.VolumeSource)
+	}
+	return fmt.Sprintf("{%s : %s}", v.Name, sourceString)
+}
+
+// HumanReadableVolumeMounts returns a string representation of a list of Kubernetes volume mounts which
+// is more compact and human readable than the default string go prints.
+func HumanReadableVolumeMounts(mounts []v1.VolumeMount) string {
+	stringMounts := []string{}
+	for _, mount := range mounts {
+		stringMounts = append(stringMounts, HumanReadableVolumeMount(&mount))
+	}
+	return fmt.Sprintf("%v", stringMounts)
+}
+
+// HumanReadableVolumeMount returns a string representation of a Kubernetes volume mount which is more
+// compact and human readable than the default string go prints.
+func HumanReadableVolumeMount(m *v1.VolumeMount) string {
+	return fmt.Sprintf("{%s : %s}", m.Name, m.MountPath)
+}
+
+/*
+ * Fully-implemented test for matching Volumes and VolumeMounts
+ */
+
+// VolumesSpec is a struct which includes a list of Kubernetes volumes as well as additional
+// metadata about the volume list for better identification during tests.
+type VolumesSpec struct {
+	// Moniker is a name given to the list to help identify it
+	Moniker string
+	Volumes []v1.Volume
+}
+
+// MountsSpec is a struct which includes a list of Kubernetes volume mounts as well as additional
+// metadata about the volume mount list for better identification during tests.
+type MountsSpec struct {
+	// Moniker is a name given to the list to help identify it
+	Moniker string
+	Mounts  []v1.VolumeMount
+}
+
+// VolumesAndMountsTestDefinition defines which volumes and mounts to test and what those values
+// should be. The test is intended to be defined with VolumesSpec defined as a pod's volumes, and
+// the list of MountsSpec items defined as the volume mounts from every container in the pod.
+type VolumesAndMountsTestDefinition struct {
+	VolumesSpec     *VolumesSpec
+	MountsSpecItems []*MountsSpec
+}
+
+// TestMountsMatchVolumes tests two things:
+// (1) That each volume mount in each every MountsSpec has a corresponding volume to source it
+//     in the VolumesSpec
+// (2) That there are no extraneous volumes defined in the VolumesSpec that do not have a
+//     corresponding volume mount in any of the MountsSpec items
+func (d *VolumesAndMountsTestDefinition) TestMountsMatchVolumes(t *testing.T) {
+	// Doing these O(n) conversions allows subsequent O(1) lookups in the map
```

Comment:
I did some more analysis here and realized that I made a mistake. Creating the map is n * (n log n) on average, so it is actually worse than doing (n^2 +n)/2 to look through the list over and over (on average). Likely it's not really pertinent for small numbers, but this could come into play for when we do tests with quite a few individual OSD mounts.